src/key_store/mod.rs

```diff
@@ -0,0 +1,222 @@
+//!     0,
+//!     0,
+//!     ManagedGroupCallbacks::default(),
+//! );
+//!
+//! let alice_group = ManagedGroup::new(
+//!     &key_store,
+//!     &managed_group_config,
+//!     GroupId::from_slice(b"Test Group"),
+//!     &alice_key_package.hash(),
+//! )
+//! .unwrap();
+//! ```
+//!
+//! # Future Work
+//!
+//! A more detailed roadmap can be found in issue #337, but generally the plan is to
+//! * move ownership of all `KeyPackageBundle` instances into the `KeyStore` and
+//!   change the API from `take_` to `get_`
+//! * keep track of which key material is (still) in use and where, as well as if it expires and when
+//! * extend the API to allow for deletion of unused key material, expired or otherwise
+//! * add a persistence layer
+
+use std::{
+    collections::HashMap,
+    sync::{RwLock, RwLockReadGuard},
+};
+
+use crate::{
+    ciphersuite::{CiphersuiteName, SignaturePublicKey, SignatureScheme},
+    credentials::{Credential, CredentialBundle, CredentialType},
+    extensions::Extension,
+    key_packages::{KeyPackage, KeyPackageBundle},
+};
+
+pub mod errors;
+
+#[cfg(test)]
+mod test_key_store;
+
+pub use errors::KeyStoreError;
+
+#[derive(Debug, Default)]
+/// The `KeyStore` contains `CredentialBundle`s and `KeyPackageBundle`s and
+/// makes them available via information in the corresponding `Credential` and
+/// `KeyPackage` instances.
+pub struct KeyStore {
+    // Map from signature public keys to credential bundles
+    credential_bundles: RwLock<HashMap<SignaturePublicKey, CredentialBundle>>,
+    init_key_package_bundles: RwLock<HashMap<Vec<u8>, KeyPackageBundle>>,
+}
+
+/// This guard struct for a `CredentialBundle` implements `Deref`, such that the
+/// underlying `RwLock<CredentialBundle>` can be obtained for read or write
+/// access to the credential.
+pub struct CBGuard<'a> {
+    cbs: RwLockReadGuard<'a, HashMap<SignaturePublicKey, CredentialBundle>>,
+    index: &'a SignaturePublicKey,
+}
+
+use std::ops::Deref;
+
+impl<'b> Deref for CBGuard<'b> {
+    type Target = CredentialBundle;
+
+    fn deref(&self) -> &CredentialBundle {
+        // We can unwrap here, as we checked if the entry is present before
+        // creating the guard. Also, since we hold a read lock on the `HashMap`,
+        // the entry can't have been removed in the meantime.
+        self.cbs.get(self.index).unwrap()
+    }
+}
+
+impl KeyStore {
+    /// Retrieve a `KeyPackageBundle` from the key store given the hash of the
+    /// corresponding `KeyPackage`. This removes the `KeyPackageBundle` from the
+    /// store. Returns an error if no `KeyPackageBundle` can be found
+    /// corresponding to the given `KeyPackage` hash.
+    pub fn take_key_package_bundle(&self, kp_hash: &[u8]) -> Option<KeyPackageBundle> {
+        // We unwrap here, because the two functions claiming write locks (this
+        // one and `generate_key_package`) only hold the lock very briefly and
+        // should not panic during that period.
```

Comment:
Hm, "should not panic" is very optimistic... We should handle all the potential locks in these functions somehow. Or add it as a todo. But this has huge potential to create bugs that are impossible to debug.

Reply:
Yes, I wasn't sure what to write here and I agree that "should not panic" is very optimistic. The way I understand poisoning is that something is considered "poisoned" if a piece of code panics while holding a write lock. This means that as long as we thoroughly check what is going on while we hold a write lock, we should be ok. It's a fine line, but I would argue that it's doable, since the only thing we (currently) do is taking the write lock, inserting or removing a value and then immediately dropping it again. Of course, this is dangerous, because adding new code could break this pattern. That being said, I still believe that if we're careful here, unwrapping should be fine.